### DISCUSSION TITLE
Reset all colors?

### DISCUSSION THREAD

**Discussion by dpc** [jj-vcs/jj#5098] [Q&A]
I'm very particular about colors and generally don't like outputs of cli tools to be too colorful. I'd like to reset all (seemingly) hardcoded color settings and only apply a handful of carefully prepared colors settings I like.

**Comment by yuja** [comment]
There's an idea to insert some indirection to switch the default color tables, but nothing is implemented yet.
https://github.com/martinvonz/jj/issues/3594

**Comment by martinvonz** [comment]
If you just want to find the colors to override, try `jj config list --include-defaults colors`.

**Reply by dpc** [reply]
Huh. I wish `jj config list --include-defaults colors | sed 's/"[a-zA-Z ]*"$/{ fg = "default", bg = "default", bold = false }/'` worked, but the way the output sometimes list the color key, and sometimes particular attributes makes it more difficult. 

And also generally I have troubles figuring out how the color overriding is supposed to work.

**Reply by martinvonz** [reply]
See https://martinvonz.github.io/jj/latest/config/#custom-colors-and-styles for the rules, although there's not that much detail about which rules override which. `jj --color=debug <command>` may also be useful.

**Reply by dpc** [reply]
The problem now is that the way compound overrides are working is unclear to me.

Does any of:

```
# 200 lines of:
"working_copy author" = { fg = "default", bg = "default", bold = false }
# etc.

# my own little settings (I basically want the prefix bold, and current one with some colors):
"prefix" = { fg = "default", bg = "default", bold = true }
# "working_copy change_id" = { fg = "blue", bg = "default", bold = false }
# "working_copy commit_id" = { fg = "green", bg = "default", bold = false }
"working_copy change_id" = "bright blue"
"working_copy commit_id" = "bright magenta"
```

supposed to work? Because it doesn't seem to.

**Reply by yuja** [reply]
Something like this might work:
```
jj config list --include-defaults colors -T'name.remove_suffix(".fg").remove_suffix(".bold") ++ "\n"' | sort -u
```

This issue will be fixed in the next release. `{ fg = .. }` will be printed as an inline table.

**Reply by dpc** [reply]
Basically if I set both `working_copy change_id` and `change_id`, then .. what exactly? :D . And should setting `working_copy change_id prefix` work? Are there any "priorities"?

**Reply by yuja** [reply]
iirc, `working_copy change_id` is prioritized over `change_id` (because the former has longer matching sequence.) `working_copy change_id prefix` will be overlaid on top of `working_copy change_id`.

**Reply by dpc** [reply]
I see, thanks. Well, something with 2-len matching sequence is overwritting `working_copy change_id` because if I change it to `"working_copy change_id prefix" = "bright blue"`, it takes effect, just only on the prefix :D. :shrug:

**Reply by dpc** [reply]
I don't think this system is very good as is. :D

**Reply by yuja** [reply]
Hmm, the actual merging logic appears to be a bit different. Something like `innermost match > longer match`?
https://github.com/martinvonz/jj/blob/main/cli/src/formatter.rs#L319-L347

(so we can't delete default colors by overriding them as `label = { fg = "default", .. }`? `label = {}` might work in the current development version, though.)

**Reply by dpc** [reply]
My intuition would be to make it an explicitly sorted (explicit priority? maybe just order in the config?) list of rules or something.

**Reply by yuja** [reply]
> maybe just order in the config?

It wouldn't work if user just wanted to override a couple of styles.
```toml
# default
"commit_id" = "blue"  # priority=1
"working_copy commit_id" = "bright blue"  # priority=2
# user
"commit_id" = "green"  # priority=3 (has higher priority than "working_copy commit_id")
```

I think the problem is that it's difficult to add fully-customized color table which can be merged with the default one, rather than the merging logic itself.
